float32[] pos
float32[] f_length
bool has_lock
